ads/google/a4a/utils.js

```diff
@@ -0,0 +1,317 @@
+ * @returns {boolean}  Whether Google Ads should attempt to render via the A4A
+ *   pathway.
+ */
+export function isGoogleAdsA4AValidEnvironment(win) {
+  const supportsNativeCrypto = win.crypto &&
+      (win.crypto.subtle || win.crypto.webkitSubtle);
+  // Note: Theoretically, isProxyOrigin is the right way to do this, b/c it
+  // will be kept up to date with known proxies.  However, it doesn't seem to
+  // be compatible with loading the example files from localhost.  To hack
+  // around that, just say that we're A4A eligible if we're in local dev
+  // mode, regardless of origin path.
+  return supportsNativeCrypto &&
+      (isProxyOrigin(win.location) || getMode().localDev);
+}
+
+/**
+ * @param {!AmpA4A} a4a
+ * @param {string} baseUrl
+ * @param {number} startTime
+ * @param {number} slotNumber
+ * @param {!Array<!QueryParameter>} queryParams
+ * @param {!Array<!QueryParameter>} unboundedQueryParams
+ * @return {!Promise<string>}
+ */
+export function googleAdUrl(
+    a4a, baseUrl, startTime, slotNumber, queryParams, unboundedQueryParams) {
+  const referrerPromise = viewerFor(a4a.getWin()).getReferrerUrl();
+  return getAdCid(a4a).then(clientId => referrerPromise.then(referrer =>
+      buildAdUrl(
+          a4a, baseUrl, startTime, slotNumber, queryParams,
+          unboundedQueryParams, clientId, referrer)));
+}
+
+
+/**
+ * @param {string} str
+ * @return {!Uint8Array}
+ * @visibleForTesting
+ */
+export function base64ToByteArray(str) {
+  const bytesAsString = atob(str);
+  const len = bytesAsString.length;
+  const bytes = new Uint8Array(len);
+  for (let i = 0; i < len; i++) {
+    bytes[i] = bytesAsString.charCodeAt(i);
+  }
+  return bytes;
+}
+
+/**
+ * @param {!ArrayBuffer} creative
+ * @param {!Headers} responseHeaders
+ * @return {!Promise<!AdResponseDef>}
+ */
+export function extractGoogleAdCreativeAndSignature(
+    creative, responseHeaders) {
+  let signature = null;
+  try {
+    if (responseHeaders.has(AMP_SIGNATURE_HEADER)) {
+      signature = base64ToByteArray(responseHeaders.get(AMP_SIGNATURE_HEADER));
+    }
+  } finally {
+    return Promise.resolve({creative, signature});
+  }
+}
+
+/**
+ * @param {!AmpA4A} a4a
+ * @param {string} baseUrl
+ * @param {number} startTime
+ * @param {number} slotNumber
+ * @param {!Array<!QueryParameter>} queryParams
+ * @param {!Array<!QueryParameter>} unboundedQueryParams
+ * @param {(string|undefined)} clientId
+ * @param {string} referrer
+ * @return {string}
+ */
+function buildAdUrl(
+    a4a, baseUrl, startTime, slotNumber, queryParams, unboundedQueryParams,
+    clientId, referrer) {
+  const global = a4a.getWin();
+  const documentInfo = documentInfoFor(global);
+  if (!global.gaGlobal) {
+    // Read by GPT for GA/GPT integration.
+    global.gaGlobal = {
+      vid: clientId,
+      hid: documentInfo.pageViewId,
+    };
+  }
+  const viewport = viewportFor(global);
+  const slotRect = viewport.getLayoutRect(a4a.element);
```

Comment:
So you're saying that we should avoid the `Viewport#getLayoutRect` because it's heavyweight?  @bobcassels and @taymonbeal worked on this parameter code and I don't know the rationale for calling both of those.  Maybe one of them can comment on whether the `#getLayoutRect` call is necessary?